xdsl/utils/lexer.py

```diff
@@ -122,3 +137,382 @@ def print_with_context(self, msg: str | None = None) -> str:
     def __repr__(self):
         return "{}[{}:{}](text='{}')".format(self.__class__.__name__,
                                              self.start, self.end, self.text)
+
+
+@dataclass(frozen=True, repr=False)
+class StringLiteral(Span):
+
+    def __post_init__(self):
+        if len(self) < 2 or self.text[0] != '"' or self.text[-1] != '"':
+            raise ParseError(self, "Invalid string literal!")
+
+    @classmethod
+    def from_span(cls, span: Span | None) -> StringLiteral | None:
+        """
+        Convert a normal span into a StringLiteral, to facilitate parsing.
+
+        If argument is None, returns None.
+        """
+        if span is None:
+            return None
+        return cls(span.start, span.end, span.input)
+
+    @property
+    def string_contents(self):
+        # TODO: is this a hack-job?
```

Comment:
IMO yes, could we transition to a hand-rolled one? Feels really weird to use `ast` for this. I have a suspicion it wouldn't be much slower, and it would definitely facilitate control over behaviour

Reply:
Yes, we will see if we should change this once I start writing on the parser itself.